flink-clients/src/main/java/org/apache/flink/client/program/DetachedEnvironment.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.client.program;
+
+import org.apache.flink.api.common.InvalidProgramException;
+import org.apache.flink.api.common.JobExecutionResult;
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.api.common.JobSubmissionResult;
+import org.apache.flink.api.common.Plan;
+import org.apache.flink.optimizer.plan.FlinkPlan;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Execution Environment for remote execution with the Client in detached mode.
+ */
+public class DetachedEnvironment extends ContextEnvironment {
+
+	/** Keeps track of the program plan for the Client to access. */
+	private FlinkPlan detachedPlan;
+
+	private static final Logger LOG = LoggerFactory.getLogger(DetachedEnvironment.class);
+
+	public DetachedEnvironment(Client remoteConnection, List<URL> jarFiles, List<URL> classpaths, ClassLoader userCodeClassLoader) {
+		super(remoteConnection, jarFiles, classpaths, userCodeClassLoader);
+	}
+
+	@Override
+	public JobExecutionResult execute(String jobName) throws Exception {
+		Plan p = createProgramPlan(jobName);
+		setDetachedPlan(Client.getOptimizedPlan(client.compiler, p, getParallelism()));
+		LOG.warn("Job was executed in detached mode, the results will be available on completion.");
+		this.lastJobExecutionResult = DetachedJobExecutionResult.INSTANCE;
+		return this.lastJobExecutionResult;
+	}
+
+	public void setDetachedPlan(FlinkPlan plan) {
+		if (detachedPlan == null) {
+			detachedPlan = plan;
+		} else {
+			throw new InvalidProgramException(DetachedJobExecutionResult.DETACHED_MESSAGE +
+					DetachedJobExecutionResult.EXECUTE_TWICE_MESSAGE);
+		}
+	}
+
+	/**
+	 * Finishes this Context Environment's execution by explicitly running the plan constructed.
+	 */
+	public JobSubmissionResult finalizeExecute() throws ProgramInvocationException {
```

Comment:
I'm not sure I understand. `execute()` should default to `execute(null)`, no? 
We don't really want to allow the user to trigger execution explicitly, which also makes me realize this should be a package private method.